File: split_array_largest_sum.py
#User function Template for python3

class Solution:
    def isvalid(self,arr,mid,n,k):
        left=k
        i=0
        while i<n and left>0:
            left-=1
            currsum=0
            while i<n and left<n-i:
                if arr[i]>mid:
                    return False
                currsum+=arr[i]
                if currsum>mid:
                    break
                i+=1
        if i<n:
            return False
        return True    
    def splitArray(self, arr, N, K):
        # code here 
        l=0
        h=sum(arr)
        ans=0
        while l<=h:
            mid=int((l+h)/2)
            if self.isvalid(arr,mid,N,K):
                ans=mid
                h=mid-1
            else:
                l=mid+1
        return ans        

#{ 
 # Driver Code Starts
#Initial Template for Python 3

if __name__ == '__main__': 
    t = int (input ())
    for _ in range (t):
        N,K=map(int,input().split())
        arr=list(map(int,input().split()))
        
        ob = Solution()
        print(ob.splitArray(arr,N,K))
# } Driver Code Ends

File: split_array_largest_sum.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends

class Solution {
  public:
    bool isvalid(int a[],long long int mid,int n,int k){
        int currSum=0,left=k,i=0;
        while(i<n && left>0){
            currSum=0;
            left--;
            while(left<n-i && i<n){
                if(a[i]>mid) return false;
                currSum+=a[i];
                if(currSum>mid){
                    break;
                }
                i++;
            }
        }
        if(i<n) return false;
        return true;
    }
    int splitArray(int arr[] ,int N, int K) {
        // code here
        long long int sum=0;
        for(int i=0;i<N;i++) sum+=arr[i];
        long long int l=0,h=sum,ans;
        while(l<=h){
            long long int mid=(l+h)/2;
            if(isvalid(arr,mid,N,K)){
                ans=mid;
                h=mid-1;
            }else{
                l=mid+1;
            }
        }
        return ans;
    }
};


//{ Driver Code Starts.

int main() {
    int t;
    cin >> t;
    while (t--) {
        int N, K;
        
        cin>>N>>K;
        int arr[N];
        
        for(int i=0 ; i<N ; i++)
            cin>>arr[i];

        Solution ob;
        cout<<ob.splitArray(arr,N,K);
        cout<<"\n";
    }
    return 0;
}
// } Driver Code Ends